# I. Разминка numpy (1 балл)

In [ ]:
import numpy as np

## 1. Парсеки

Напишите функцию, которая принимает на вход:
- np.array - набор расстояний от Земли до звёзд (в световых годах)
- k - количество значений, которое нам нужно

Выведите топ-k самых удаленных звёзд от нас в порядке убывания разрастания, расстояния выведите в парсеках (1 парсек приблизительно равен 3.262 световым годам)

In [ ]:
def find_top_k_longest_distance_stars(stars, k):
  sorted_stars = np.sort(stars)[::-1]
  sorted_stars = sorted_stars / 3.262
  return sorted_stars[:k:]

In [ ]:
# Просто пример, можно придумать свой
print(find_top_k_longest_distance_stars(np.array([4.37, 8.6, 4.24, 4.37, 5.96, 7.8, 8.6, 7.9]), 4))

[2.63641937 2.63641937 2.4218271  2.39117106]


## 2. Оцениваем размер галактики

Напишите функцию, которая принимает на вход:
- np.array - набор масс звёзд в галактике
- k - количество максимальных по массе звёзд

Функция должна вывести два числа - первое - сумму масс первых k звёзд по массе, второе - сумму масс остальных звёзд.

In [ ]:
def count_k_masses(stars, k):
  sorted_stars = np.sort(stars)[::-1]
  sum_heaviest = np.sum(sorted_stars[:k])
  sum_other = np.sum(sorted_stars) - sum_heaviest
  return sum_heaviest, sum_other

In [ ]:
# Просто пример, можно придумать свой
print(count_k_masses(np.array([150, 200, 300, 250, 400, 100, 200]), 2))

(700, 900)


# II. Помогаем Карелии

Смотреть на просто бюджет

In [ ]:
import pandas as pd

## 0. Загрузи датасет в этот блокнот (0.5 балла)

1.   List item
2.   List item



Открой гугл-диск курса, разархивируй zip-архив и собери нужные наборы данных из архива в единый датасет при помощи pandas.

In [ ]:
def prep_df(df):
    df.columns = list(df.loc["Показатель"])
    df = df.drop("Показатель", axis=0)
    return df
# Эта функция была написана Николаем в качестве примера для Полины Тихоновой, поэтому надеюсь норм, что я ее использую

files = [
    "2010.xlsx",
    "2011.xlsx",
    "2012.XLS",
    "2013.xls",
    "2014.XLS",
    "2015.xls",
    "2016.xlsx",
    "2017.xlsx",
    "2018.xlsx",
    "2019.xlsx",
    "2020.xlsx"
]

files_bad = [
    "2012.XLS",
    "2013.xls",
    "2014.XLS",
    "2015.xls"
]
# Файлы формата .xls плохие - в них есть ненужный лист, при их обработке не нужно его добавлять в датафреймы

df = []

for item in files:
  if item in files_bad:
    file_1 = pd.read_excel(item, sheet_name = 1)
    file_2 = pd.read_excel(item, sheet_name = 2)
  else:
    file_1 = pd.read_excel(item)
    file_2 = pd.read_excel(item, sheet_name = 1)
  df.append(file_1)
  df.append(file_2)
# Считываем файлы: расходы добавляем к расходам, доходы - к доходам

for i in range(0, 22):
  if i <= 3 or i == 6 or i == 8 or i == 10:
    df[i] = df[i].iloc[:, [0,-1]]
  elif i == 4:
    df[i] = df[i].iloc[:, [0,4]]
  elif i == 5 or i == 7 or i == 9:
    df[i] = df[i].iloc[:, [0,6]]
  elif i == 11:
    df[i] = df[i].iloc[:, [0,7]]
  elif i == 12:
    df[i] = df[i].iloc[:, [1,5]]
  elif i % 2 == 1:
    df[i] = df[i].iloc[:, [0,3]]
  else:
    df[i] = df[i].iloc[:, [1,5]]
  df[i] = df[i].transpose()
  # print(df[i])
#Сейчас у нас есть 22 таблички, состоящие из двух столбцов: наименование статьи и численное значение

for i in range (22):
   df[i].iloc[0] = df[i].iloc[0].str.lower()
   df[i].index = ['Показатель', 2010 + ((i+1)//2 - i % 2)]
   if i < 12:
     df[i] = df[i].drop(df[i].columns[0], axis=1)
   elif i == 16:
     df[i] = df[i].drop(df[i].columns[0], axis=1)
   elif i == 20:
     df[i] = df[i].drop(df[i].columns[0], axis=1)
   elif i % 2 == 1:
     df[i] = df[i].drop(df[i].columns[0], axis=1)
   else:
     df[i] = df[i].drop(df[i].columns[1], axis=1)
   df[i].columns = [str(i) for i in range(len(df[i].columns))]
# Для удобства сделали все названия столбцов lowercase и изменили названия строк, удалили ненужный 1й столбец


for i in [0, 2, 5, 7]:
  df[i] = df[i].drop(df[i].columns[0], axis=1)
df[5] = df[5].drop(df[5].columns[0:2], axis=1)
df[7] = df[7].drop(df[7].columns[0:2], axis=1)
df[20] = df[20].drop(df[20].columns[0], axis=1)
for i in range(4, 12):
  df[i] = df[i].drop(df[i].columns[1], axis=1)
  df[i].columns = [str(i) for i in range(len(df[i].columns))]
# Избавились от ненужных колонок

for i in range(0, 12, 2):
  df[i].iat[0, 0] = 'Доходы бюджета'
  df[i+1].iat[0, 0] = 'Расходы бюджета'
for i in range(12, 22, 2):
  df[i].iat[0, 0] = 'Расходы бюджета'
  df[i+1].iat[0, 0] = 'Доходы бюджета'
df[7].iat[0, 4] = 'оплата труда и начисления на выплаты по оплате труда'
# Переименовали первый столбец во что-то нормальное, пофиксили столбец "ъ"

for i in range(22):
    df[i] = (df[i]).loc[:, ~prep_df(df[i]).columns.duplicated()]
# Удалили дубликаты столбцов

for i in range(22):
  df[i] = df[i].iloc[1:, :]
# Удалили первую строку, в которой нет данных, из каждой таблицы

df_1 = [df[0], df[2], df[4], df[6], df[8], df[10], df[13], df[15], df[17], df[19], df[21]]
df_2 = [df[1], df[3], df[5], df[7], df[9], df[11], df[12], df[14], df[16], df[18], df[20]]

combined_income = pd.concat(df_1, ignore_index=True)
combined_income.index = [2010 + i for i in range(11)]
combined_expenses = pd.concat(df_2, ignore_index=True)
combined_expenses.index = [2010 + i for i in range(11)]
combined = [combined_income, combined_expenses]
# 2 таблицы: одна с расходами, другая с доходами

merged = pd.concat(combined, axis = 1)
display(merged)



Доходы бюджета налоговые и неналоговые доходы  \
2010  24873441175.450001                 15322132232.73   
2011      27556466680.93                 18202689345.18   
2012  26893154749.540001             17570424288.029999   
2013  23984242603.040001                 15173763025.93   
2014      27750206180.02                  15956764902.0   
2015      28390875283.93             17808163620.080002   
2016  32546447594.290001             19654412574.419998   
2017  34385355496.239998             20129196349.220001   
2018  47268601337.099998                 26604349315.43   
2019  51272761265.220001             27763981934.919998   
2020  60597379486.300003                 27131141282.57   

     налоги на прибыль, доходы налог на прибыль организаций  \
2010         8613483384.459999                 3980212142.4   
2011            11334538970.16                6262562069.63   
2012        10813732699.530001                 5279887784.7   
2013             8101885886.87                2200021502.01   
2014             8916328469.16                1837840442.62   
2015         9962889434.440001                3114014266.32   
2016            11518070774.48                4280042099.75   
2017        11700807336.309999                4022683492.19   
2018               16900358608                8128882485.27   
2019        17336881098.560001                8162340492.33   
2020        15165765187.879999                5058407816.68   

     налог на прибыль организаций, зачисляемый в бюджеты бюджетной системы российской федерации по соответствующим ставкам  \
2010                                       3980212142.4                                                                      
2011                                      6262562069.63                                                                      
2012                                       5279887784.7                                                                      
2013                                      2200021502.01                                                                      
2014                                      1837840442.62                                                                      
2015                                      3114014266.32                                                                      
2016                                      4280042099.75                                                                      
2017                                      4022683492.19                                                                      
2018                                      8128882485.27                                                                      
2019                                      8162340492.33                                                                      
2020                                      5058407816.68                                                                      

     налог на прибыль организаций, зачисляемый в бюджеты субъектов российской федерации  \
2010                                      3980213558.01                                   
2011                                      6262518004.63                                   
2012                                       5279864760.7                                   
2013                                      2200021502.01                                   
2014                                      1837840442.62                                   
2015                                      2589450837.68                                   
2016                                                NaN                                   
2017                                                NaN                                   
2018                                                NaN                                   
2019                                                NaN                                   
2020                          

## 1. Предобработка датасета (2 балла)

- изучи датасет
- очисти данные в нём (есть ли в нём пропуски и выбросы? что ты можешь с ними сделать? сделай это.)

In [ ]:
import warnings
import numpy as np
from scipy.stats import zscore

In [ ]:
def NaNs(column):
    median = column.median()
    new_column = column.fillna(median)
    return new_column
# Заменяет NaN на медиану столбца

def outliers(column):
    z_scores = zscore(column)
    median = column.median()
    outliers = np.abs(z_scores) > 3
    column.loc[outliers] = median
    return column
# Заменяем выбросы на медиану столбца

warnings.simplefilter(action='ignore', category=FutureWarning)
# Код выводил очень много предупреждений, поэтому я их отключил

df = merged

df = df.loc[:, df.isnull().sum() <= 3]
# Уберем все столбцы, в которых количество пропусков больше, чем 3

df = df.apply(NaNs)
df = df.apply(outliers)
# Все выбросы и пропуски заменяем на медиану столбца.

pd.set_option('display.float_format', '{:.2f}'.format)
display(df['Доходы бюджета'], df['Расходы бюджета'])
# Вручную убедились, что в столбцах с общими доходами и расходами ничего не поменялось

display(df)
df_1 = df


2010   24873441175.45
2011   27556466680.93
2012   26893154749.54
2013   23984242603.04
2014   27750206180.02
2015   28390875283.93
2016   32546447594.29
2017   34385355496.24
2018   47268601337.10
2019   51272761265.22
2020   60597379486.30
Name: Доходы бюджета, dtype: float64

2010   25641875414.57
2011   27631246903.81
2012   29140749888.23
2013   28754420604.42
2014   31039085086.34
2015   31356669427.82
2016   34508364905.09
2017   36810091699.46
2018   43076921835.03
2019   50268395809.96
2020   67616189896.19
Name: Расходы бюджета, dtype: float64

Доходы бюджета  налоговые и неналоговые доходы  \
2010  24873441175.45                  15322132232.73   
2011  27556466680.93                  18202689345.18   
2012  26893154749.54                  17570424288.03   
2013  23984242603.04                  15173763025.93   
2014  27750206180.02                  15956764902.00   
2015  28390875283.93                  17808163620.08   
2016  32546447594.29                  19654412574.42   
2017  34385355496.24                  20129196349.22   
2018  47268601337.10                  26604349315.43   
2019  51272761265.22                  27763981934.92   
2020  60597379486.30                  27131141282.57   

      налоги на прибыль, доходы  налог на прибыль организаций  \
2010              8613483384.46                 3980212142.40   
2011             11334538970.16                 6262562069.63   
2012             10813732699.53                 5279887784.70   
2013              8101885886.87                 2200021502.01   
2014              8916328469.16                 1837840442.62   
2015              9962889434.44                 3114014266.32   
2016             11518070774.48                 4280042099.75   
2017             11700807336.31                 4022683492.19   
2018             16900358608.00                 8128882485.27   
2019             17336881098.56                 8162340492.33   
2020             15165765187.88                 5058407816.68   

      налог на прибыль организаций, зачисляемый в бюджеты бюджетной системы российской федерации по соответствующим ставкам  \
2010                                      3980212142.40                                                                       
2011                                      6262562069.63                                                                       
2012                                      5279887784.70                                                                       
2013                                      2200021502.01                                                                       
2014                                      1837840442.62                                                                       
2015                                      3114014266.32                                                                       
2016                                      4280042099.75                                                                       
2017                                      4022683492.19                                                                       
2018                                      8128882485.27                                                                       
2019                                      8162340492.33                                                                       
2020                                      5058407816.68                                                                       

      налог на доходы физических лиц  \
2010                   4633271242.06   
2011                   5071976900.53   
2012                   5533844914.83   
2013                   5901864384.86   
2014                   7078488026.54   
2015                   6848875168.12   
2016                   7238028674.73   
2017                   7678123844.12   
2018                   8771476122.73   
2019                   9174540606.23   
2020                  10107357371.20   

      налоги на товары (работы, услуги), реализуемые на территории российской федерации  \
2010                                      2022770319.58                                   
2011                                      2137298390.60                                   
2012                                      2522931839.55                                   
2013                                      2339682080.95                                   
2014                                      2101870504.68                                   


## 2. Найди параметры, которые влияют на бюджет (3.5 балла)

Поэкспериментируй с параметрами, найди 3 параметра, которые сильно связаны с значением бюджета Карелии за год (тут есть две цифры - расходы бюджета ИТОГО и доходы бюджета ИТОГО - искать параметры надо к обоим на основе **предыдущего года**) (при помощи коэффициента корреляции). Настоятельно рекомендуем поиграть с разными параметрами  - попробовать линейные комбинации, некоторые функции от параметров (log, возведение в степень, etc).

Оставь весь код экспериментов в этом блоке, но в конце опиши, какие параметры в итоге ты берёшь.

In [ ]:
df = df_1
df['след_год_доходы'] = df['Доходы бюджета'].shift(-1)
df['след_год_расходы'] = df['Расходы бюджета'].shift(-1)
# Так как нас интересует связь со доходов / расходов с параметрами предыдущего года

# display(df['след_год_доходы'], df[след_год_расходы])

correlations_income = df.corr()['след_год_доходы'].sort_values(ascending=False)
correlations_expenses = df.corr()['след_год_расходы'].sort_values(ascending=False)
# Вычисляем корреляцию параметров с доходами и расходами.
# Я смотрю на корреляцию всего со всем (то есть, в том числе рассматриваю влияние статей расходов на доходы следующего года, и наоборот)


print("Наибольшая корреляция с доходами:", end = "\n")
display(correlations_income.dropna().head(10))
print("Наибольшая антикорреляция с доходами:", end = "\n")
display(correlations_income.dropna().tail(10))
print("Наибольшая корреляция с расходами:", end = "\n")
display(correlations_expenses.dropna().head(10))
print("Наибольшая антикорреляция с расходами:", end = "\n")
display(correlations_expenses.dropna().tail(10))
# Топ параметров с наибольшей (положительной или отрицательной) корреляцией

# Логарифмы
log_columns = {}

for column in df.columns:
        if (df[column] > 0).all():
            log_columns[f'log_{column}'] = np.log(df[column])  # Логарифмируем столбцы

log_df = pd.DataFrame(log_columns, index=df.index)
log_df = pd.concat([df['Доходы бюджета'], df['Расходы бюджета'], log_df], axis=1)

correlations_income_log = log_df.corr()['Доходы бюджета'].sort_values(ascending=False)
correlations_expenses_log = log_df.corr()['Расходы бюджета'].sort_values(ascending=False)
# Вычисляем корреляцию логпараметров с доходами и расходами.

print("Наибольшая корреляция с доходами (логарифмы параметров):")
display(correlations_income_log.dropna().head(10))
print("Наибольшая антикорреляция с доходами (логарифмы параметров):")
display(correlations_income_log.dropna().tail(10))
print("Наибольшая корреляция с расходами (логарифмы параметров):")
display(correlations_expenses_log.dropna().head(10))
print("Наибольшая антикорреляция с расходами (логарифмы параметров):")
display(correlations_expenses_log.dropna().tail(10))
# Топ параметров с наибольшей (положительной или отрицательной) корреляцией





Наибольшая корреляция с доходами:


след_год_доходы                                                                                                                 1.000000
другие вопросы в области социальной политики                                                                                    0.974378
спорт высших достижений                                                                                                         0.970207
другие общегосударственные вопросы                                                                                              0.969895
след_год_расходы                                                                                                                0.969448
налог, взимаемый с налогоплательщиков, выбравших в качестве объекта налогообложения доходы, уменьшенные на величину расходов    0.967482
социальная политика                                                                                                             0.962587
Расходы бюджета                                                                                                                 0.961983
другие вопросы в области образования                                                                                            0.960204
функционирование высшего должностного лица субъекта российской федерации и муниципального образования                           0.959212
Name: след_год_доходы, dtype: float64

Наибольшая антикорреляция с доходами:


доходы, получаемые в виде арендной платы за земельные участки, государственная собственность на которые не разграничена и которые расположены в границах городских округов, а также средства от продажи права на заключение договоров аренды указанных земельных участков           -0.632834
доходы, получаемые в виде арендной платы за земельные участки, государственная собственность на которые не разграничена, а также средства от продажи права на заключение договоров аренды указанных земельных участков                                                              -0.632834
общее образование                                                                                                                                                                                                                                                                   -0.687061
доходы от сдачи в аренду имущества, составляющего государственную (муниципальную) казну (за исключением земельных участков)                                                                                                                                                         -0.698995
доходы от сдачи в аренду имущества, составляющего казну субъекта российской федерации (за исключением земельных участков)                                                                                                                                                           -0.698995
доходы, получаемые в виде арендной платы за земли после разграничения государственной собственности на землю, а также средства от продажи права на заключение договоров аренды указанных земельных участков (за исключением земельных участков бюджетных и автономных учреждений)   -0.700999
денежные взыскания (штрафы) за нарушение правил перевозки крупногабаритных и тяжеловесных грузов по автомобильным дорогам общего пользования регионального или межмуниципального значения                                                                                           -0.703491
прочие безвозмездные поступления                                                                                                                                                                                                                                                    -0.721757
здравоохранение                                                                                                                                                                                                                                                                     -0.726380
государственная пошлина за государственную регистрацию межрегиональных, региональных и местных общественных объединений, отделений общественных объединений, а также за государственную регистрацию изменений их учредительных документов                                           -0.821968
Name: след_год_доходы, dtype: float64

Наибольшая корреляция с расходами:


след_год_расходы                                                                       1.000000
другие общегосударственные вопросы                                                     0.990527
судебная система                                                                       0.990212
Расходы бюджета                                                                        0.985777
другие вопросы в области культуры, кинематографии                                      0.973992
след_год_доходы                                                                        0.969448
безвозмездные поступления                                                              0.966166
спорт высших достижений                                                                0.965546
безвозмездные поступления от других бюджетов бюджетной системы российской федерации    0.964187
Доходы бюджета                                                                         0.959379
Name: след_год_расходы, dtype: float64

Наибольшая антикорреляция с расходами:


доходы, получаемые в виде арендной либо иной платы за передачу в возмездное пользование государственного и муниципального имущества (за исключением имущества бюджетных и автономных учреждений, а также имущества государственных и муниципальных унитарных предприятий, в том числе казенных)   -0.563921
общее образование                                                                                                                                                                                                                                                                                 -0.570441
денежные взыскания (штрафы) за нарушение правил перевозки крупногабаритных и тяжеловесных грузов по автомобильным дорогам общего пользования регионального или межмуниципального значения                                                                                                         -0.602653
доходы, получаемые в виде арендной платы за земли после разграничения государственной собственности на землю, а также средства от продажи права на заключение договоров аренды указанных земельных участков (за исключением земельных участков бюджетных и автономных учреждений)                 -0.616471
здравоохранение                                                                                                                                                                                                                                                                                   -0.625202
прочие безвозмездные поступления в бюджеты субъектов российской федерации                                                                                                                                                                                                                         -0.640427
прочие безвозмездные поступления                                                                                                                                                                                                                                                                  -0.664762
доходы от сдачи в аренду имущества, составляющего государственную (муниципальную) казну (за исключением земельных участков)                                                                                                                                                                       -0.674501
доходы от сдачи в аренду имущества, составляющего казну субъекта российской федерации (за исключением земельных участков)                                                                                                                                                                         -0.674501
государственная пошлина за государственную регистрацию межрегиональных, региональных и местных общественных объединений, отделений общественных объединений, а также за государственную регистрацию изменений их учредительных документов                                                         -0.676247
Name: след_год_расходы, dtype: float64

Наибольшая корреляция с доходами (логарифмы параметров):


Доходы бюджета                                                                             1.000000
log_Доходы бюджета                                                                         0.994027
log_безвозмездные поступления                                                              0.989303
log_безвозмездные поступления от других бюджетов бюджетной системы российской федерации    0.988760
log_Расходы бюджета                                                                        0.980605
log_другие вопросы в области культуры, кинематографии                                      0.974446
Расходы бюджета                                                                            0.971636
log_другие общегосударственные вопросы                                                     0.971592
log_судебная система                                                                       0.971524
log_органы юстиции                                                                         0.964882
Name: Доходы бюджета, dtype: float64

Наибольшая антикорреляция с доходами (логарифмы параметров):


log_плата за выбросы загрязняющих веществ в атмосферный воздух стационарными объектами                                                                                                                                                                                                                -0.630592
log_государственная пошлина за государственную регистрацию политических партий и региональных отделений политических партий                                                                                                                                                                           -0.686999
log_доходы, получаемые в виде арендной либо иной платы за передачу в возмездное пользование государственного и муниципального имущества (за исключением имущества бюджетных и автономных учреждений, а также имущества государственных и муниципальных унитарных предприятий, в том числе казенных)   -0.737373
log_проценты, полученные от предоставления бюджетных кредитов внутри страны                                                                                                                                                                                                                           -0.801374
log_проценты, полученные от предоставления бюджетных кредитов внутри страны за счет средств бюджетов субъектов российской федерации                                                                                                                                                                   -0.801374
log_государственная пошлина за государственную регистрацию межрегиональных, региональных и местных общественных объединений, отделений общественных объединений, а также за государственную регистрацию изменений их учредительных документов                                                         -0.817458
log_прочие безвозмездные поступления в бюджеты субъектов российской федерации                                                                                                                                                                                                                         -0.857199
log_прочие безвозмездные поступления                                                                                                                                                                                                                                                                  -0.871765
log_доходы от сдачи в аренду имущества, составляющего государственную (муниципальную) казну (за исключением земельных участков)                                                                                                                                                                       -0.893103
log_доходы от сдачи в аренду имущества, составляющего казну субъекта российской федерации (за исключением земельных участков)                                                                                                                                                                         -0.893103
Name: Доходы бюджета, dtype: float64

Наибольшая корреляция с расходами (логарифмы параметров):


Расходы бюджета                                                                            1.000000
log_Расходы бюджета                                                                        0.990209
log_обеспечение пожарной безопасности                                                      0.976772
Доходы бюджета                                                                             0.971636
log_безвозмездные поступления                                                              0.967859
log_судебная система                                                                       0.965321
log_безвозмездные поступления от других бюджетов бюджетной системы российской федерации    0.964134
log_сельское хозяйство и рыболовство                                                       0.960245
log_органы юстиции                                                                         0.955352
log_другие общегосударственные вопросы                                                     0.952813
Name: Расходы бюджета, dtype: float64

Наибольшая антикорреляция с расходами (логарифмы параметров):


log_обслуживание государственного (муниципального) долга                                                                                                                                                                                                                                              -0.575103
log_государственная пошлина за государственную регистрацию политических партий и региональных отделений политических партий                                                                                                                                                                           -0.594918
log_доходы, получаемые в виде арендной либо иной платы за передачу в возмездное пользование государственного и муниципального имущества (за исключением имущества бюджетных и автономных учреждений, а также имущества государственных и муниципальных унитарных предприятий, в том числе казенных)   -0.688694
log_государственная пошлина за государственную регистрацию межрегиональных, региональных и местных общественных объединений, отделений общественных объединений, а также за государственную регистрацию изменений их учредительных документов                                                         -0.847357
log_проценты, полученные от предоставления бюджетных кредитов внутри страны                                                                                                                                                                                                                           -0.850705
log_проценты, полученные от предоставления бюджетных кредитов внутри страны за счет средств бюджетов субъектов российской федерации                                                                                                                                                                   -0.850705
log_прочие безвозмездные поступления                                                                                                                                                                                                                                                                  -0.861942
log_доходы от сдачи в аренду имущества, составляющего государственную (муниципальную) казну (за исключением земельных участков)                                                                                                                                                                       -0.904981
log_доходы от сдачи в аренду имущества, составляющего казну субъекта российской федерации (за исключением земельных участков)                                                                                                                                                                         -0.904981
log_прочие безвозмездные поступления в бюджеты субъектов российской федерации                                                                                                                                                                                                                         -0.915994
Name: Расходы бюджета, dtype: float64

<font color="violet"> Тут надо написать параметры, которые ты выбрал </font>

---



<font color="violet"> Выбранные параметры

<font color="violet"> Доходы:

<font color="blue"> log_безвозмездные поступления	(0.989303)

<font color="blue"> log_безвозмездные поступления от других бюджетов бюджетной системы российской федерации (0.988760)

 <font color="blue">спорт высших достижений (0.970207)

<font color="violet"> Расходы:

<font color="blue">другие общегосударственные вопросы	(0.990527)

<font color="blue">судебная система	(0.990212)

<font color="blue">другие вопросы в области культуры, кинематографии	(0.973992)

# 3. Графики (1 балл)

Построй графики с выбранными параметрами и бюджетом Карелии. Что ты на них видишь?

In [ ]:
years = []
for i in range(2010, 2021):
  years.append(i)


In [ ]:
import plotly.express as px
df = df_1

graph_1 = (df['безвозмездные поступления']) / (df['безвозмездные поступления'].shift(1))
graph_2 = (df['безвозмездные поступления от других бюджетов бюджетной системы российской федерации']) / (df['безвозмездные поступления от других бюджетов бюджетной системы российской федерации'].shift(1))
graph_3 = df['спорт высших достижений'] / df['спорт высших достижений'].shift(1)
graph_4 = df['Доходы бюджета'] / df['Доходы бюджета'].shift(1)

plot_data = pd.DataFrame({
    'Год': [years[i] for i in range(11)],
    'Безвозмездные поступления': graph_1,
    'Безвозмездные поступления от других бюджетов': graph_2,
    'Спорт высших достижений': graph_3,
    'Доходы бюджета': graph_4,
})

figure = px.line(
    plot_data,
    x='Год',
    y=[
    'Безвозмездные поступления',
    'Безвозмездные поступления от других бюджетов',
    'Спорт высших достижений',
    'Доходы бюджета'
    ],
    labels={
        'value': 'Годовой рост',
        'variable': 'Категория'
    },
    title='Годовой рост доходов и бюджетных статей'
)

figure.show()

In [ ]:
import plotly.express as px
df = df_1

graph_1 = df['другие общегосударственные вопросы'] / df['другие общегосударственные вопросы'].shift(1)
graph_2 = df['судебная система'] / df['судебная система'].shift(1)
graph_3 = df['другие вопросы в области культуры, кинематографии'] / df['другие вопросы в области культуры, кинематографии'].shift(1)
graph_4 = df['Расходы бюджета'] / df['Расходы бюджета'].shift(1)

plot_data = pd.DataFrame({
    'Год': [years[i] for i in range(11)],
    'Другие общегосударственные вопросы': exp_1,
    'Судебная система': exp_2,
    'Другие вопросы в области культуры, кинематографии': exp_3,
    'Расходы бюджета': exp_4,
})

figure = px.line(
    plot_data,
    x='Год',
    y=[
    'Другие общегосударственные вопросы',
    'Судебная система',
    'Другие вопросы в области культуры, кинематографии',
    'Расходы бюджета'
    ],
    labels={
        'value': 'Годовой рост',
        'variable': 'Категория'
    },
    title='Годовой рост расходов и бюджетных статей'
)

figure.show()

## 4. Анализ параметров (2 балла)

На основе графиков и самих параметров, расскажи:

- Построй гипотезы, наблюдаешь ли ты казуацию или здесь просто совпадение?
- Предложи гипотезы, как проверить, нет ли ошибок в твоих выводах
- Проанализируй взаимосвязь с параметрами между собой. Являются ли они сильноскореллированными? Имеет ли смысл использовать их всех вместе или достаточно одного?

Что можно было заметить:

0) Корреляция каждой переменной с ней же самой 1

1) Во-первых, доходы и расходы бюджета сильно скоррелированы: логично, потому что обычно бюджет стараются формировать без дефицита + чем больше у региона доходов, тем выше будут расходы

2) Во-вторых, очень явная корреляция между доходами бюджета и выплатами, эти доходы формирующими: "безвозмездными поступлениями" и
"безвозмездными поступлениями от других бюджетов бюджетной системы российской федерации", это может свидетельствовать о том, что бюджет Карелии определяется во многом выплатами из федерального центра.

3) Что касается "спорта высших достижений", это скорее совпадение, чем каузация, хотя может иметь место третий фактор, о котором я не подумал. Кроме того, у этой статьи есть выделяющийся рост в более, чем 2 раза, в 2017 году - на графике хорошо виден этот пик, возможно, это выброс, который мы не почистили

4) Видим какую-то корреляцию между расходами бюджета и расходами на судебную систему и культуру. Возможно, это связано с тем, что эти расходы непервостепенные => будут меняться из года в год, но тогда лучше посмотреть на их корреляцию еще и с доходами

5) Точно можно сказать, что мы рассмотрели 2 взаимосвязанных показателя: "безвозмездными поступлениями" и
"безвозмездными поступлениями от других бюджетов бюджетной системы российской федерации": на графике доходов это красная и синии линии, и они в точности повторяют друг друга. Достаточно использовать только один показатель из двух.

Гипотезы, как проверить наличие ошибок в выводах:

1) Самый простой способ, который решит МНОЖЕСТВО проблем, возникающих при анализе бюджета и построении линейной модели, - взять больше данных. Тяжело что-то анализировать всего на 11 наблюдениях, из которых мы возможно выкинули часть важных параметров (одно и то же называется по-разному, из-за этого появляеся много NaN'ов, которыые потом удаляются)

2) Проанализировать обстоятельства, которые могли повлиять на показатели и составляющие доходов / расходов


## 5. Линейная модель - BONUS TASK! - 2 балла.

Бонусное задание - попробуй построить линейную модель бюджета на основе выбранных параметров выше и (возможно) других параметров из датасета. Как думаешь, она состоятельная? Как бы ты проверял эту модель на корректность?

Вспоминаем Машинное Обучение - 1 и используем линейную регрессию (так как хотим предсказывать целевую переменную)

Для доходов:

In [ ]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import train_test_split, KFold, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error as mse


x_objects = df[
    [
    'безвозмездные поступления',
    'безвозмездные поступления от других бюджетов бюджетной системы российской федерации',
    'другие вопросы в области культуры, кинематографии',
    'спорт высших достижений',
    'другие вопросы в области социальной политики'
    ]
] # Признаки
y_objects = df['Доходы бюджета']  # Целевая переменная


scaler_x = StandardScaler()
scaler_y = StandardScaler()
x_scaled = scaler_x.fit_transform(x_objects)
y_scaled = scaler_y.fit_transform(y_objects.values.reshape(-1, 1)).flatten()
# Масштабируем признаки и целевую переменную

ridge_model = Ridge(alpha=1.0)
ridge_model.fit(x_scaled, y_scaled)
# Обучаем модель

y_pred = ridge_model.predict(x_scaled)

mean_mse = mse(y_scaled, y_pred)
weights = ridge_model.coef_
print(f'MSE модели: {mean_mse:.4f}')
print('Веса модели:')
for feature, weight in zip(x_columns, weights):
    print(f'{feature}: {weight:.4f}')

MSE модели: 0.0230
Веса модели:
безвозмездные поступления: 0.2943
безвозмездные поступления от других бюджетов бюджетной системы российской федерации: 0.2858
другие вопросы в области культуры, кинематографии: 0.2172
спорт высших достижений: 0.0217
другие вопросы в области социальной политики: 0.1643


Для расходов:

In [ ]:
x_objects = df[
  [
  'другие общегосударственные вопросы',
  'судебная система',
  'другие вопросы в области культуры, кинематографии',
  'безвозмездные поступления'
  ]
] # Признаки
y_objects = df['Расходы бюджета']  # Целевая переменная


scaler_x = StandardScaler()
scaler_y = StandardScaler()
x_scaled = scaler_x.fit_transform(x_objects)
y_scaled = scaler_y.fit_transform(y_objects.values.reshape(-1, 1)).flatten()
# Масштабируем признаки и целевую переменную

ridge_model = Ridge(alpha=1.0)
ridge_model.fit(x_scaled, y_scaled)
# Обучаем модель

y_pred = ridge_model.predict(x_scaled)

mean_mse = mse(y_scaled, y_pred)
weights = ridge_model.coef_
print(f'MSE модели: {mean_mse:.4f}')
print('Веса модели:')
for feature, weight in zip(x_columns, weights):
    print(f'{feature}: {weight:.4f}')

MSE модели: 0.0245
Веса модели:
безвозмездные поступления: 0.2230
безвозмездные поступления от других бюджетов бюджетной системы российской федерации: 0.2353
другие вопросы в области культуры, кинематографии: 0.1237
спорт высших достижений: 0.3862


Попытались построить модель, но состоятельной, она, скорее всего, не будет: бессмысленно пытаться предсказать что-то на таком маленьком объеме данных. По-хорошему, следовало бы использовать k-fold кросс-валидацию и другие методы для более объективной оценки модели, но на наборе из 11 точек любой инструмент будет избыточным, поэтому я их и не использую.

Для проверки корректности модели можно:

1) Всё же использовать кросс-валидацию

2) Попытаться что-то визуализировать, например, построить графики

3) Смотреть на MSE, MAE, MAPE

4) Тщательно подбирать вручную параметр регуляризации